### PR TITLE
Refactored Search Tabs

### PR DESCRIPTION
**Pull request by ThoWagen**
The initial purpose of this PR was to also display the result count of the active search tab.
Besides this the SearchTab view helper was simplified, a wrong doc comment was fixed and jQuery was removed from resultscount.js.

### CODE REVIEW

File: themes/bootstrap3/js/resultcount.js [hunk 1]
```diff
@@ -2,19 +2,29 @@
...
-        $.ajax({
-          url: VuFind.path + '/AJAX/JSON?method=getResultCount',
-          dataType: 'json',
-          data: {querystring: queryString, source: source},
-          success: function appendResultCount(response){
-            $this.append(' (' + response.data.total.toLocaleString() + ')');
+    document.querySelectorAll('ul.nav-tabs [data-show-counts] a').forEach((tab) => {
+      function loadCount(url) {
+        let source = tab.dataset.source;
+        let xhttp = new XMLHttpRequest();
```
Comment (demiankatz) [inline themes/bootstrap3/js/resultcount.js]: Could this be simplified using the fetch api?

File: themes/bootstrap3/js/resultcount.js [hunk 2]
```diff
@@ -2,19 +2,29 @@
...
+        let params = new URLSearchParams({
+          method: 'getResultCount',
+          querystring: url,
+          source: source
         });
+        xhttp.open('GET', VuFind.path + '/AJAX/JSON?' + params.toString());
+        xhttp.setRequestHeader('Content-type', 'application/json');
+        xhttp.send();
+      }
+      if (tab.getAttribute('href')) {
```
Comment (demiankatz) [inline themes/bootstrap3/js/resultcount.js]: Could this be simplified to `loadCount(tab.getAttribute('href') || tab.dataset.searchUrl)` (maybe in combination with an early return from loadCount if a falsy value is passed in)?

File: module/VuFind/src/VuFind/View/Helper/Root/SearchTabs.php [hunk 3]
```diff
@@ -331,90 +303,41 @@ protected function remapBasicSearch(
...
-     * @param array  $filters        Tab filters
-     * @param string $permissionName Name of a permissionrule
+     * @param string $class   Search class ID
+     * @param array  $filters Tab filters
      *
-     * @return array
+     * @return string
      */
-    protected function createHomeTab($id, $class, $label, $filters, $permissionName)
+    protected function getHomeTabUrl($class, $filters)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/SearchTabs.php:314]: I realize that this is a pre-existing problem, but the comment at the top of this method is duplicated in getAdvancedTabUrl(), and I'm not sure if it really makes sense in either context.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen, this does seem to make things significantly simpler and eliminate a lot of pointless redundancy.

I left a few comments below.

One other general question: in most cases, won't we have the count of results for the current tab already embedded on the page somewhere? Would it make sense to add logic to try to find and duplicate that value if possible before resorting to repeating the whole search in an AJAX call?

File: themes/bootstrap3/js/resultcount.js (hunk 1 above)
Comment (ThoWagen) [reply themes/bootstrap3/js/resultcount.js]: Done

File: module/VuFind/src/VuFind/View/Helper/Root/SearchTabs.php (hunk 3 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Root/SearchTabs.php:314]: I guess that's my fault that I didn't update the comments. Done.

**Comment by ThoWagen** [comment]
When the tabs are included in searchTabs.phtml we can use the '$results' object to set the count directly. But it doesn't always exist e.g. on "search home". And we can't use this for the combined tab since the it's '$results' object has always 'resultTotel=0'.

File: themes/bootstrap3/js/resultcount.js (hunk 2 above)
Comment (ThoWagen) [reply themes/bootstrap3/js/resultcount.js]: Done

**Comment by demiankatz** [comment]
> When the tabs are included in searchTabs.phtml we can use the '$results' object to set the count directly. But it doesn't always exist e.g. on "search home". And we can't use this for the combined tab since the it's '$results' object has always 'resultTotel=0'.

All good points. But is it perhaps worth using the number from the results object when rendering the current selected tab, if it is set? (We could also include an edge case to check that it's greater than zero if that remains a concern after work on #3310 is completed). As long as we have the AJAX method available as a fallback for the edge cases where the data is not immediately available, I think it's worth using the available data in those situations where we can, since I think that is going to eliminate a significant number of unnecessary HTTP requests and searches.

**Comment by ThoWagen** [comment]
Now, the result count is set directly if possible.

I also set `$resultTotal` to -1 by default since 0 could actually be a valid value for it. The count is only shown if its greater or equal to 0. This automatically implements your suggestion in #3310.

File: module/VuFind/src/VuFind/Search/Base/Results.php [hunk 4]
```diff
@@ -309,7 +309,7 @@ public function performAndProcessSearch()
     {
         // Initialize variables to defaults (to ensure they don't stay null
         // and cause unnecessary repeat processing):
-        $this->resultTotal = 0;
+        $this->resultTotal = -1;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/Base/Results.php:313]: Maybe worth adding a comment here to explain that we're choosing -1 to flag "not processed yet" separately from "no results"? Adjusting the doc comment for the getResultTotal() method might also be in order.

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php [hunk 5]
```diff
@@ -135,7 +135,7 @@ public function process($results)
     public function getVisFacets()
     {
         // Don't bother processing if the result set is empty:
-        if ($this->searchObject->getResultTotal() == 0) {
+        if ($this->searchObject->getResultTotal() <= 0) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php:138]: If you're trying to catch all the comparisons between getResultTotal() and 0, it looks like one got missed in \VuFindApi\Formatter\FacetFormatter.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen -- this seems to be working correctly for me (and all tests are passing). Just a couple of small suggestions (mostly documentation-related).

I've also requested a review from @EreMaijala in case he can think of any reason _not_ to switch to the -1 default, since he's been working with low-level search code more recently than I have and might think of something I'm overlooking. Please feel free to wait for his input before acting on my suggestions, in case he suggests more significant changes.

File: module/VuFind/src/VuFind/Search/Base/Results.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Search/Base/Results.php:313]: If we already have `null` for "not processed yet", should -1 rather be "not available"? Or am I missing something?

Comment (demiankatz) [reply module/VuFind/src/VuFind/Search/Base/Results.php:313]: Yes, you're correct, "not available" is a much better term.

Comment (ThoWagen) [reply module/VuFind/src/VuFind/Search/Base/Results.php:313]: Done

File: module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php (hunk 5 above)
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Recommend/PubDateVisAjax.php:138]: Fixed

**Review by demiankatz — APPROVED** [review]
Looks good to me now, and all tests are passing. Thanks, @ThoWagen. As discussed, I have also closed #3310 since it is superseded by work here.

**Pull request merged**